### PR TITLE
Add simd math intrinsics and gather/scatter

### PR DESCRIPTION
**Pull request by gnzlbg**
This PR adds simd math intrinsics for floating-point vectors (sqrt, sin, cos, pow, exp, log, fma, abs, etc.) and the generic simd gather/scatter intrinsics.

### CODE REVIEW

**Comment by gnzlbg** [comment]
Still needs to be cleaned up and need to add run-time tests, but I'd like initial feedback on the general approach.

**Comment by gnzlbg** [comment]
@petrochenkov all tests are there

**Comment by alexcrichton** [comment]
Nice!

For the math intrinsics though, those can all be defined natively with `#[link_name]`, right? If that's the case should we prefer that instead?

**Comment by gnzlbg** [comment]
@eddyb for the gather/scatter intrinsics, I've kept it simple for now, but I'll try to add the llvm functions that generate the function names to rustllvm and use those instead. Some of the tests failed with llvm 3.9 because 

```
Intrinsic name not mangled correctly for type arguments! Should be: llvm.masked.scatter.v2f32
[01:00:08] void (<2 x float>, <2 x float*>, i32, <2 x i1>)* @llvm.masked.scatter.v2f32.v2p0f32
[01:00:08] LLVM ERROR: Broken function found, compilation aborted!
```

which is surprising to me because I did not know that LLVM intrinsics name "style" changed at some point in time. I hope that it doesn't change again in a way that forces us to rename all `link_llvm_intrinsics` in `stdsimd` or forces us to support multiple link names.

**Comment by alexcrichton** [comment]
@gnzlbg oh it's fine if stdsimd only supports the latest LLVM currently, and so most of the tests are fine to just ignore on non-HEAD LLVM. 

If the math intrinsics can be defined with `link_llvm_intrinsics` can they be left out to explicitly define in stdsimd?

**Comment by alexcrichton** [comment]
(I'd prefer to only put in the compiler what *must* be there for one reason or another)

**Comment by gnzlbg** [comment]
@alexcrichton I've made the intrinsics "approximative". Given that it was a one line change, I am still unsure of whether it wouldn't be better to just handle both cases in rustc.

**Comment by alexcrichton** [comment]
Ah ok I see! Yeah if these end up looking like the "fast math" intrinsics we have it seems fine by me to land. In that case...

@bors: r+

File: src/test/run-pass/simd-intrinsic-float-math.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    fn simd_fsin<T>(x: T) -> T;
+    fn simd_fcos<T>(x: T) -> T;
+    fn simd_ceil<T>(x: T) -> T;
+    fn simd_fexp<T>(x: T) -> T;
+    fn simd_fexp2<T>(x: T) -> T;
+    fn simd_floor<T>(x: T) -> T;
+    fn simd_fma<T>(x: T, y: T, z: T) -> T;
+    fn simd_flog<T>(x: T) -> T;
+    fn simd_flog10<T>(x: T) -> T;
+    fn simd_flog2<T>(x: T) -> T;
```
Comment (kennytm) [inline src/test/run-pass/simd-intrinsic-float-math.rs]:
```
[01:29:38] ---- [run-pass] run-pass/simd-intrinsic-float-math.rs stdout ----
[01:29:38] 	
[01:29:38] error: compilation failed!
[01:29:38] status: exit code: 101
[01:29:38] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/simd-intrinsic-float-math.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--target=arm-linux-androideabi" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.stage2-arm-linux-androideabi" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/arm-linux-androideabi/native/rust-test-helpers" "-Clinker=/android/ndk/arm-14/bin/arm-linux-androideabi-clang" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.stage2-arm-linux-androideabi.aux"
[01:29:38] stdout:
[01:29:38] ------------------------------------------
[01:29:38] 
[01:29:38] ------------------------------------------
[01:29:38] stderr:
[01:29:38] ------------------------------------------
[01:29:38] error: linking with `/android/ndk/arm-14/bin/arm-linux-androideabi-clang` failed: exit code: 1
[01:29:38]   |
[01:29:38]   = note: "/android/ndk/arm-14/bin/arm-linux-androideabi-clang" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-Wl,--allow-multiple-definition" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math0-317d481089b8c8fe83113de504472633.rs.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math2-317d481089b8c8fe83113de504472633.rs.rcgu.o" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.stage2-arm-linux-androideabi" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.crate.allocator.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "-L" "/checkout/obj/build/arm-linux-androideabi/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.stage2-arm-linux-androideabi.aux" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib" "-Wl,--start-group" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib" "-l" "std-dd7e1cb952f01bfe" "-Wl,--end-group" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib/libcompiler_builtins-93fbfd71748200b0.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "log" "-l" "gcc" "-l" "gcc" "-l" "c" "-l" "m" "-Wl,-rpath,$ORIGIN/../../stage2/lib/rustlib/arm-linux-androideabi/lib" "-Wl,-rpath,/checkout/obj/lib/rustlib/arm-linux-androideabi/lib" "-Wl,--enable-new-dtags"
[01:29:38]   = note: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs.rcgu.o:simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs:function simd_intrinsic_float_math::main::h3154db08c8edf39b: error: undefined reference to 'log2f'
[01:29:38]           /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs.rcgu.o:simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs:function simd_intrinsic_float_math::main::h3154db08c8edf39b: error: undefined reference to 'log2f'
[01:29:38]           /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs.rcgu.o:simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs:function simd_intrinsic_float_math::main::h3154db08c8edf39b: error: undefined reference to 'log2f'
[01:29:38]           /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/simd-intrinsic-float-math.simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs.rcgu.o:simd_intrinsic_float_math1-317d481089b8c8fe83113de504472633.rs:function simd_intrinsic_float_math::main::h3154db08c8edf39b: error: undefined reference to 'log2f'
[01:29:38]           clang50: error: linker command failed with exit code 1 (use -v to see invocation)
[01:29:38]           
[01:29:38] 
[01:29:38] error: aborting due to previous error
[01:29:38] 
[01:29:38] 
[01:29:38] ------------------------------------------
[01:29:38] 
[01:29:38] thread '[run-pass] run-pass/simd-intrinsic-float-math.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3033:9
[01:29:38] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by kennytm** [comment]
@bors r-

The new test failed on `arm-android` (`error: undefined reference to 'log2f'`). See https://github.com/rust-lang/rust/pull/50775#issuecomment-389299570 for detailed log.

**Comment by gnzlbg** [comment]
So it appears that this failed to link but only on `arm-android`: https://travis-ci.org/rust-lang/rust/builds/379359939

@kennytm would it be possible to re-start the stopped build bots? It would be useful to me to know if this fails somewhere else.  For example, it works on `armhf-gnu` and `dist-arm-android`, so... maybe it is because it requires `hf` ?

**Comment by gnzlbg** [comment]
So this does indeed only trigger on that single build bot. Googling for "undefined reference log2f android" reveals: https://github.com/rust-lang/rust/issues/31119 which suggests that this is fixed by using the NDK API 18. However, it seems that that particular build bot is using the ndk API 14 ? https://github.com/rust-lang/rust/blob/master/src/ci/docker/arm-android/Dockerfile#L8

I don't know what to do here but maybe an `// ignore arm-android` is the way to go until we update the ndk version on CI, if ever. Thoughts ?

**Comment by pietroalbini** [comment]
Ping from triage @alexcrichton! This PR needs your review.

**Comment by alexcrichton** [comment]
@bors: r+

er oops, sorry this fell off my radar!

**Pull request merged**